reconcile/external_resources/manager.py

```diff
@@ -0,0 +1,423 @@
+                    reconcile_status = self.impl.get_resource_reconcile_status(key)
+                    if reconcile_status != ReconcileStatus.IN_PROGRESS:
+                        state.ts = datetime.now(timezone.utc)
+                        if reconcile_status == ReconcileStatus.SUCCESS:
+                            state.status = ResourceStatus.CREATED
+                        elif reconcile_status == ReconcileStatus.ERROR:
+                            state.status = ResourceStatus.ERROR
+                        logging.info("Resource Status updated to " + state.status)
+            except StateNotFoundError:
+                state = ExternalResourceState(
+                    key=key,
+                    ts=datetime.now(timezone.utc),
+                    status=ResourceStatus.NOT_EXISTS,
+                )
+            except Exception:
+                raise
+
+            resource = self.build_external_resource(spec)
+            if self.resource_needs_reconciliation(state=state, resource=resource):
+                logging.info("External Resource needs reconcile " + spec.identifier)
+                image = self.get_provider_image(spec)
+                input = self.serialize_resource_input(resource)
+                self.impl.reconcile_resource(
+                    key=key,
+                    image=image,
+                    input=input,
+                    action=Action.APPLY,
+                    dry_run=self.dry_run,
+                )
+
+                state.ts = datetime.now(timezone.utc)
+                state.status = ResourceStatus.IN_PROGRESS
+                state.resource_hash = resource.digest()
+                state.reconcile_data = ExternalResourceReconcileData(
+                    key=state.key,
+                    image=image,
+                    input=input,
+                )
+                self.state_mgr.set_external_resource_state(key, state)
+
+    def handle_deleted_resources(
+        self,
+        specs: Iterable[ExternalResourceSpec],
+    ) -> None:
+        pass
+        state_resource_keys = self.state_mgr.get_all_resource_keys()
+        desired_resource_keys = [ExternalResourceKey.from_spec(s) for s in specs]
+        diff = set(state_resource_keys) - set(desired_resource_keys)
+        for key in diff:
+            state = self.state_mgr.get_external_resource_state(key)
+            if state.status == ResourceStatus.DELETE_IN_PROGRESS:
+                reconcile_status = self.impl.get_resource_reconcile_status(key)
+                if reconcile_status != ReconcileStatus.IN_PROGRESS:
+                    state.ts = datetime.now(timezone.utc)
+                    if reconcile_status == ReconcileStatus.SUCCESS:
+                        state.status = ResourceStatus.DELETED
+                    elif reconcile_status == ReconcileStatus.ERROR:
+                        state.status = ResourceStatus.ERROR
+                    logging.info("Updating resource status to " + state.status)
```

Comment:
similar here, worth extracting method, btw I'm thinking `handle_desired_resources` and `handle_deleted_resources` are called together, can we simplify it so one method handle iterating and pair desired state and current state.

```python
    def handle_resources(
        self,
        specs: Iterable[ExternalResourceSpec],
    ) -> None:
        spec_by_key = {ExternalResourceKey.from_spec(spec): spec for spec in specs}
        diff_result = diff_mappings(
            self.state_mgr.index,
            spec_by_key,
            lambda _c, _d: True,
        )
        
        for key, spec in diff_result.add.items():
            handle_add_resource(key, spec)
        
        for key, pair in diff_result.identical.items():
            handle_identical_resource(key, pair.current, pair.desired)
        
        for key, current in diff_result.delete.items():
            handle_delete_resource(key, current)
```

```python

```

Reply:
This also changed, I can not use your suggestions as to get the specs for the deleted resources I need to do a table scan.